lib/Transforms/IRUtils.cpp

```diff
@@ -0,0 +1,211 @@
+    // Create new variable to hold `val`.
+    auto newVar = rewriter.create<P4HIR::VariableOp>(
+        val.getLoc(), P4HIR::ReferenceType::get(val.getType()), varName);
+
+    mlir::OpBuilder::InsertionGuard guard(rewriter);
+
+    // Insert a new read in all escaping blocks and replace uses of `val` in that block.
+    for (mlir::Block *block : escapingBlocks) {
+        rewriter.setInsertionPointToStart(block);
+        auto newVal = rewriter.create<P4HIR::ReadOp>(val.getLoc(), newVar);
+        rewriter.replaceUsesWithIf(val, newVal, [block](mlir::OpOperand &use) {
+            return use.getOwner()->getBlock() == block;
+        });
+    }
+
+    // Assign the new variable after `val`'s definition.
+    rewriter.setInsertionPointAfterValue(val);
+    rewriter.create<P4HIR::AssignOp>(val.getLoc(), val, newVar);
+}
+
+// Replace uses of `op` with a copy in all escaping blocks.
+void copyOp(mlir::RewriterBase &rewriter, mlir::Operation *op, const BlockSet &escapingBlocks) {
+    mlir::OpBuilder::InsertionGuard guard(rewriter);
+    for (mlir::Block *block : escapingBlocks) {
+        rewriter.setInsertionPointToStart(block);
+        auto newOp = rewriter.clone(*op);
+        rewriter.replaceOpUsesWithinBlock(op, newOp->getResults(), block);
+    }
+}
+
+};  // namespace
+
+void IRUtils::adjustBlockUses(mlir::RewriterBase &rewriter, mlir::Block *block) {
+    // Iterate in reverse so we process uses before their definition in this block.
+    // This is needed to properly handle operands after `copyOp` calls.
+    for (mlir::Operation &op : llvm::make_early_inc_range(llvm::reverse(*block))) {
+        if (mlir::isa<P4HIR::VariableOp>(op)) {
+            // Promote variables to a parser locals.
+            rewriter.moveOpBefore(&op, rewriter.getInsertionBlock(), rewriter.getInsertionPoint());
+        } else if (mlir::isa<P4HIR::ConstOp, P4HIR::StructExtractRefOp, P4HIR::ArrayElementRefOp,
+                             P4HIR::SliceRefOp>(op)) {
+            // Copy RefOps and constants.
+            auto escapingBlocks = getEscapingBlocks(block, op.getResult(0));
+            copyOp(rewriter, &op, escapingBlocks);
+        } else {
+            // Promote values to variables.
+            for (auto val : op.getResults()) {
+                auto escapingBlocks = getEscapingBlocks(block, val);
+
+                if (!escapingBlocks.empty()) promoteValToVar(rewriter, val, escapingBlocks);
+            }
+        }
+    }
+}
+
+P4HIR::ParserStateOp IRUtils::createSubState(mlir::RewriterBase &rewriter,
+                                             P4HIR::ParserStateOp state, llvm::StringRef suffix) {
+    auto parser = state->getParentOfType<P4HIR::ParserOp>();
+    std::string basename = (llvm::Twine(state.getSymName()) + "_" + suffix).str();
+    std::string name = basename;
+
+    size_t counter = 0;
+    while (parser.lookupSymbol(name)) name = basename + "_" + std::to_string(counter++);
```

Comment:
Use `mlir::SymbolTable::generateSymbolName`. There are examples of usage in `P4C/translate.cpp`